File: Utils/src/main/java/com/sheryv/util/VersionUtils.java
package com.sheryv.util;

import org.apache.commons.lang3.StringUtils;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;

public class VersionUtils {
  
  public static Version loadVersionByModuleName(String moduleName) {
    InputStream resourceAsStream = VersionUtils.class.getClassLoader().getResourceAsStream(moduleName + ".txt");
    long ms = 0;
    try (BufferedReader reader = new BufferedReader(new InputStreamReader(resourceAsStream))) {
      StringBuilder builder = new StringBuilder();
      String version = reader.readLine();
      String time = reader.readLine();
      if (StringUtils.isNotBlank(time)) {
        ms = Long.parseLong(time);
      }
      return new Version(version, ms);
    } catch (Exception e) {
      e.printStackTrace();
    }
    return new Version("", ms);
  }
}

File: Utils/src/main/kotlin/com/sheryv/util/VersionUtils.kt
package com.sheryv.util

import org.apache.commons.lang3.StringUtils
import java.io.BufferedReader
import java.io.InputStreamReader
import java.time.Instant

object VersionUtils {
  fun loadVersionByModuleName(moduleName: String): Version {
    val resourceAsStream = VersionUtils::class.java.getClassLoader().getResourceAsStream("$moduleName.txt")
    var ms: Long = 0
    try {
      BufferedReader(InputStreamReader(resourceAsStream)).use { reader ->
        val version = reader.readLine()
        val time = reader.readLine()
        if (StringUtils.isNotBlank(time)) {
          ms = time.toLong()
        }
        return Version(version, ms)
      }
    } catch (e: Exception) {
      e.printStackTrace()
    }
    return Version("", ms)
  }
}

class Version(val version: String, val buildTimeMs: Long) {
  
  fun toTimestamp(): Instant {
    return Instant.ofEpochMilli(buildTimeMs)
  }
  
  override fun toString(): String {
    return version + " (Built at " + toTimestamp() + ')'
  }
}
